orca/requester.py

```diff
@@ -13,33 +14,124 @@
 
 def file_from_opendap(url, outdir, outfile):
     """Write to file from OPeNDAP link"""
-    urls = bisect_request(url)
+    file_extension = url.split(".")[-1].split("?")[0]
+    if file_extension != "nc":
+        outpath = to_file(url, outdir, outfile, nc=False)
+    else:
+        urls = bisect_request(url)
 
-    logger.debug(f"Downloading from {len(urls)} URL(s): {urls}")
-    dataset = open_mfdataset(urls, combine="nested", concat_dim="time")
-    outpath = to_file(dataset, outdir, outfile)
+        logger.debug(f"Downloading from {len(urls)} URL(s): {urls}")
+        if len(urls) == 1:
+            dataset = open_dataset(urls[0])
+        else:
+            dataset = open_mfdataset(urls, combine="nested", concat_dim="time")
+        outpath = to_file(dataset, outdir, outfile)
 
     return outpath
 
 
-def to_file(dataset, outdir, outfile=""):
-    """Write netcdf to generated filename"""
+def to_file(dataset, outdir, outfile="", nc=True):
+    """Write data to generated filename"""
     if not outfile:
         now = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
         outfile = f"orca-output-{now}"
 
     outpath = outdir + outfile
 
     logger.debug("Begin file write")
-    dataset.to_netcdf(outpath)
+    if nc:
+        dataset.to_netcdf(outpath)
+    else:
+        f = open(outpath, "wb")
+        f.write(requests.get(dataset).content)
+        f.close()
     logger.debug("File write complete")
 
     return outpath
 
 
 def build_opendap_url(thredds_base, filepath, targets):
-    """Construct url for OPeNDAP"""
-    return f"{thredds_base}{filepath}?{targets}"
+    """Construct url for OPeNDAP. If target variables are not given, then all variables
+    are added to the url to download the entire data file"""
+    if filepath.endswith("nc"):
+        if targets:
+            targets = targets.replace(
+                "[:]", "[]"
+            )  # These are both treated as obtaining the full range of a dimension
+            targets = fill_target_bounds(
+                f"{thredds_base}{filepath}", targets
+            )  # Ensures that entire ranges of variables are obtained
+        else:
+            targets = build_all_targets(f"{thredds_base}{filepath}")
+        return f"{thredds_base}{filepath}?{targets}"
+
+    else:  # .dds, .dds, or .ascii request
```

Comment:
Would it be cleaner to detect metadata requests in `routes.py` and redirect to THREDDS immediately before the request even gets down to this level?